### PR TITLE
Update container handling with charliecloud

### PR DESCRIPTION
**Pull request by nschan**
Here are changes to the charliecloud driver.

- `charliecloud.writeFake` is true by default (!)

- if `charliecloud.writeFake` is true and `$CH_IMAGE_STORAGE` is set, images will go into `$CH_IMAGE_STORAGE` and be run from there with --write-fake (will be run by name in that case).

- if charliecloud.writeFake is true, but `$CH_IMAGE_STORAGE` is unset, images will go into `charliecloud.cacheDir`, `$NXF_CHARLIECLOUD_CACHE` or `work/charliecloud` (order of priority) and will be run by path (since run by name only works if `$CH_IMAGE_STORAGE `is set)

- if `charliecloud.writeFake` is set to `false`, images will need to be run with `-w` to enable mounting of the workDir. In this case `$CH_IMAGE_STORAGE` will not be used to store images, since `$CH_IMAGE_STORAGE` is incompatible with `-w`. Images will instead go into `charliecloud.cacheDir`, `$NXF_CHARLIECLOUD_CACHE` or `work/charliecloud` (order of priority). If the folder used to store images is the same as `$CH_IMAGE_STORAGE`, throw an exception. Images will be run by path.

In summary:

- Give preference to `$CH_IMAGE_STORAGE` in the default case of `writeFake`

- Make some effort to avoid using `$CH_IMAGE_STORAGE` when `writeFake` is `false`.

- No longer create temporary containers for each task. 

This is the same as #5212, but without the many merge conflicts created by a failed sign-off.

@phue I have updated the documentation. I have not added a unit test for switching between `image` and `imageName` in the builder, since I do not know of a way to set an env var for the tests. I have tested this locally (i.e. running tests with the var set or not) and it produced the expected results. Is it really worth adding a way to inject fake environments into the BuilderTest to test this specific behaviour?

### CODE REVIEW

File: docs/config.md [hunk 1]
```diff
@@ -488,16 +488,18 @@ The following settings are available:
...
 The following settings are available:
 
-`charliecloud.cacheDir`
-: The directory where remote Charliecloud images are stored. When using a computing cluster it must be a shared folder accessible to all compute nodes.
-
 `charliecloud.enabled`
 : Enable Charliecloud execution (default: `false`).
 
+`charliecloud.writeFake`
+: Enable `writeFake` with charliecloud (default: `true`) This allows to run containers from storage in writeable mode, using overlayfs. `writeFake` requires `overlayfs` (Linux kernel >= 5.11). For full suppornt tempfs with xattrs in the user namespace (Linux kernel >= 6.6) is required , see [charliecloud documentation](https://hpc.github.io/charliecloud/ch-run.html#ch-run-overlay) for details.
```
Comment (phue) [inline docs/config.md:498]:
```suggestion
: Enable `writeFake` with charliecloud (default: `true`) This allows to run containers from storage in writeable mode, using overlayfs. `writeFake` requires unprivileged `overlayfs` (Linux kernel >= 5.11). For full support, tempfs with xattrs in the user namespace (Linux kernel >= 6.6) is required, see [charliecloud documentation](https://hpc.github.io/charliecloud/ch-run.html#ch-run-overlay) for details.
```

Comment (nschan) [reply docs/config.md]: I have pushed this in a new commit, I cant sign-off correctly via github.

**Comment by nschan** [comment]
> > #3566 has been fixed a while ago
> 
> How was it fixed?

Trying to catch it and retry https://github.com/nextflow-io/nextflow/blob/master/modules/nextflow/src/main/groovy/nextflow/container/CharliecloudCache.groovy#L186

**Comment by nschan** [comment]
Thanks for the helpful discussion and review @phue ! @pditommaso do you have any additional comments on this?

File: modules/nextflow/src/main/groovy/nextflow/container/CharliecloudCache.groovy [hunk 2]
```diff
@@ -130,19 +131,38 @@ class CharliecloudCache {
         if( config.pullTimeout )
             pullTimeout = config.pullTimeout as Duration
 
+        def writeFake = true as Boolean
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/container/CharliecloudCache.groovy:134]:
```suggestion
        def writeFake = true
```

**Review by pditommaso — COMMENTED** [review]
Just lest a minor comment

**Comment by nschan** [comment]
Thanks @pditommaso, I have put in the suggested change.

**Comment by pditommaso** [comment]
Nice, let's wait for the tests
